File: go/stats.go
package leanprompt

// CompressionStats — telemetry shape shared by every compression call.
type CompressionStats struct {
	InputTokens  int64
	OutputTokens int64
	// Ratio is a derived float — recomputed per language, never byte-compared.
	Ratio   float64
	Method  string
	CostUSD float64
}

func statsWithMethod(method string) CompressionStats {
	return CompressionStats{Ratio: 1.0, Method: method}
}

func statsCounted(method string, input, output int64) CompressionStats {
	ratio := 1.0
	if input > 0 {
		ratio = float64(output) / float64(input)
	}
	return CompressionStats{
		InputTokens:  input,
		OutputTokens: output,
		Ratio:        ratio,
		Method:       method,
	}
}

File: rust/src/stats.rs
//! CompressionStats — telemetry shape shared by every compression call.

#[derive(Debug, Clone, PartialEq)]
pub struct CompressionStats {
    pub input_tokens: i64,
    pub output_tokens: i64,
    /// Derived float — recomputed per language, never byte-compared.
    pub ratio: f64,
    pub method: String,
    pub cost_usd: f64,
}

impl CompressionStats {
    pub fn passthrough() -> Self {
        Self::with_method("passthrough")
    }

    pub fn with_method(method: &str) -> Self {
        CompressionStats {
            input_tokens: 0,
            output_tokens: 0,
            ratio: 1.0,
            method: method.to_string(),
            cost_usd: 0.0,
        }
    }

    pub fn counted(method: &str, input_tokens: i64, output_tokens: i64) -> Self {
        CompressionStats {
            input_tokens,
            output_tokens,
            ratio: if input_tokens > 0 {
                output_tokens as f64 / input_tokens as f64
            } else {
                1.0
            },
            method: method.to_string(),
            cost_usd: 0.0,
        }
    }
}
